File: snippets/continuation/continuation.clj
(defn find-first-odd [coll success failure]
  (if (empty? coll)
    (failure)
    (let [first-element (first coll)]
      (if (odd? first-element)
        (success first-element)
        (find-first-odd (rest coll) success failure)))))

;; (find-first-odd [2 3 4] identity (constantly nil))
;; (find-first-odd [2 2 4] identity (constantly nil))

(defn collect-odd [coll collector]
  (if (empty? coll)
    (collector nil)
    (let [first-element (first coll)]
      (if (odd? first-element)
        (collect-odd (rest coll)
                     (fn [collected]
                       (collector (conj collected first-element))))
        (collect-odd (rest coll)
                     collector)))))

;; (collect-odd [1 2 3 4] identity)

(defn subseq-with-sum [total coll success failure]
  (if-let [match (first (filter #(= % total) coll))]
    (success [match])
    (let [candidates (filter #(< % total) coll)]
      (if (empty? candidates)
        (failure)
        (let [first-candidate (first candidates)]
          (subseq-with-sum
           (- total first-candidate) (rest candidates)
           (fn [collected]
             (success (conj collected first-candidate)))
           (fn []
             (subseq-with-sum total (rest candidates)
                              success failure))))))))


;; (subseq-with-sum 10 [1 2 5 6 7] identity (constantly nil))
;; (subseq-with-sum 10 [1 2 5] identity (constantly nil))

File: snippets/continuation/continuation.py
def oddp(num):
    return num % 2 == 1

def identity(x):
    return x

def constantly(x):
    def inner():
        return x
    return inner

def find_first_odd(l, success, failure):
    if not l:
        return failure()

    first = l[0]
    if oddp(first):
        return success(first)

    return find_first_odd(l[1:], success, failure)

# print(find_first_odd([0, 2, 3, 1, 4], identity, constantly(None)))
# print(find_first_odd([0, 2], identity, constantly(None)))

# --------------------------

def collect_odd(l, success):
    if not l:
        return success([])
    first = l[0]

    if oddp(first):
        def new_success(l):
            l.append(first)
            return success(l)
        return collect_odd(l[1:], new_success)
    else:
        return collect_odd(l[1:], success)

# print(collect_odd([1, 2, 3, 4], identity))
# print(collect_odd([2, 4], identity))

# -----------------------

def subseq_with_sum(sum, l, success, failure):
    if sum == 0:
        return success([])

    candidates = list(filter(lambda x: x <= sum, l))
    if not candidates:
        return failure()

    first_candidate = candidates[0]
    rest_candidates = candidates[1:]

    def new_success(l):
        l.append(first_candidate)
        return success(l)

    def new_failure():
        return subseq_with_sum(sum, rest_candidates, success, failure)

    return subseq_with_sum(sum-first_candidate, rest_candidates, new_success, new_failure)

# print(subseq_with_sum(10, [1, 2, 5, 6, 7], identity, constantly([])))
# print(subseq_with_sum(10, [1, 2, 5, 6], identity, constantly([])))
